### ISSUE TITLE
Conflict with 'dist' and 'kern' feature in Indic scripts: kernFeatureWriter.py

### ISSUE THREAD

**Issue by rahulgajjar** [googlefonts/ufo2ft#440]
I was compiling Kannada fonts. In Kannada, there could be many below base forms with the base glyphs. When two or more non-spacing below base forms are attached to the base forms, they overlap each-other. There are two ways this problem could be solved, either making all possible combination of all below base forms and making ligatures of them. Or using kerning to give space between them. While kerning was used to give space between the below base forms, when fonts are compiled it creates one lookup where all positioning code is placed, and that lookup is referred in 'dist' and 'kern' feature blocks.

While the fonts compiled this way may work at some places, some of the softwares i.e. InDesign executes both 'dist' and 'kern' and thus making the double kerning space.

In kernFeatureWriter I noticed that there is a global variable DIST_ENABLED_SCRIPTS which is responsible for enabling 'dist' feature. But later on it adds 'kern' feature block anyway. Which is hardcoded in the beginning with self.context.todo. 

    features = frozenset(["kern", "dist"])

For testing, when I added a checkpoint ctx.tag and later on use it in _makeFeatureBlocks,

    def setContext(self, font, feaFile, compiler=None):
        ctx = super(KernFeatureWriter, self).setContext(
            font, feaFile, compiler=compiler
        )
        ctx.gdefClasses = ast.getGDEFGlyphClasses(feaFile)
        ctx.kerning = self.getKerningData(font, feaFile, self.getOrderedGlyphSet())

        feaScripts = ast.getScriptLanguageSystems(feaFile)
        ctx.scriptGroups = self._groupScriptsByTagAndDirection(feaScripts)
        for scriptCode, scriptLangSys in feaScripts.items():
            if scriptCode in DIST_ENABLED_SCRIPTS:
                ctx.tag = "dist"
            else:
                ctx.tag = "kern"

        return ctx

This way I was able to generate either 'kern' or 'dist' block. It solved the issue I had and it works in the softwares I tested.

    def _makeFeatureBlocks(self, lookups):
        features = {}
        if "kern" in self.context.todo: 
            if self.context.tag == "kern":
                kern = ast.FeatureBlock("kern")
                self._registerKernLookups(kern, lookups)
                if kern.statements:
                    features["kern"] = kern
        if "dist" in self.context.todo:
            if self.context.tag == "dist":
                dist = ast.FeatureBlock("dist")
                self._registerDistLookups(dist, lookups)
                if dist.statements:
                    features["dist"] = dist
        return features

I am not entirely sure about how all of the code works, I am not sure if this fix is proper or not. This issue arise with Devanagari as well when kerning is being used in any of the glyphs. It makes the double kerning space. I assume this would happen with all of the script which use 'dist' feature.

**Comment by khaledhosny** [comment]
> While the fonts compiled this way may work at some places, some of the softwares i.e. InDesign executes both 'dist' and 'kern' and thus making the double kerning space.

If more than one GPOS feature point to the same lookup, that lookup should be applied only once. Do you know any other applications that double apply the kerning in this case other than InDesign?

**Comment by rahulgajjar** [comment]
> If more than one GPOS feature point to the same lookup, that lookup should be applied only once. Do you know any other applications that double apply the kerning in this case other than InDesign?

Double kerning happens in Adobe Indesign 2021 and Adobe Illustrator 2021. Apparently it works fine in Adobe Photoshop 2021. Adobe text shaping engine is very messy, use HarfBuzz already! I guess it is a bug on Adobe side, don't think they are going to listen anyway. I am developing Takri Typeface, and InDesign just crashes as soon as I paste the Takri text. I have sent dozens of reports. It's frustrating. 

Thank you for your help.

**Comment by belluzj** [comment]
We also opened a bug report on Adobe's side more than a year ago: https://community.adobe.com/t5/indesign/devanagari-layouting-world-ready-paragraph-composer-applies-dflt-s-kern-on-top-of-deva-s-dist/td-p/10683539?page=1

**Comment by khaledhosny** [comment]
> Double kerning happens in Adobe Indesign 2021 and Adobe Illustrator 2021. Apparently it works fine in Adobe Photoshop 2021.

Makes sense, as [Photoshop now uses HarfBuzz](https://twitter.com/ken_lunde/status/1290744201445912576) while the other two don’t yet use it ([not by default](https://twitter.com/vinodbala/status/1285259248973213696), at least).

**Comment by anthrotype** [comment]
one way to work around this could be to split up the lookups even further and make sure the lookups registered for kern and dist do not overlap. Right now we only split by script horizontal direction.

**Comment by rahulgajjar** [comment]
So, you are saying if I have Latin glyphs and Devanagari glyphs, the kerning groups for Latin should go to 'kern' feature block and the kerning for Devanagari glyphs should go to 'dist' feature block?

This would be tricky, how can we determine which glyphs belongs to which language and thus which feature block it should go. Especially the glyphs which does not have unicode value.

This is a weird issue, for Latin glyphs, indesign does not apply double kerning. Maybe because it has Latin unicode value and it knows that it belongs to 'kern' feaure block. But, in the same font, for devanagari glyphs with unicode value it apply double kerning

**Comment by rahulgajjar** [comment]
I found out that Adobe is planning to implement Harfbuzz for adobe applications. It is available to try using scripting panel. They will provide UI options after they have tested Harfbuzz with all scripts. Finally, a good news. In Kannada, the default shaping engine was also causing another issue, when we define mark glyphs manually by using GDEF table, it was rendering text incorrectly. 

Harfbuzz workflow in InDesign:

https://community.adobe.com/t5/indesign/script-for-enabling-harfbuzz-workflow-in-indesign/td-p/11295166?page=1

**Comment by rahulgajjar** [comment]
Hi @belluzj 

I noticed another irregularity during production of Takri script font. In my font there is a manual 'dist' feature block. If I remove 'dist' feature block then 'KernFeatureWriter' generates both 'dist' and 'kern' feature block. But if there alredy 'dist' block exists, it skip making 'dist' feature block.

I am uncertain why 'dist' is removed from context.todo list even if it already checked by scruptGroups that the font needs 'dist' feature block.

Adding following block in setContext solves the issue temporary.

```
        if "dist" in self.context.scriptGroups:
            self.context.todo.add("dist")
```

```
    def setContext(self, font, feaFile, compiler=None):
        ctx = super().setContext(font, feaFile, compiler=compiler)
        ctx.gdefClasses = ast.getGDEFGlyphClasses(feaFile)
        ctx.kerning = self.getKerningData(font, feaFile, self.getOrderedGlyphSet())

        feaScripts = ast.getScriptLanguageSystems(feaFile)
        ctx.scriptGroups = self._groupScriptsByTagAndDirection(feaScripts)

        if "dist" in self.context.scriptGroups:
            self.context.todo.add("dist")

        return ctx
```

Can you please have a look at it?

Thanks

**Comment by anthrotype** [comment]
Are you using fontmake or a build script of yours that import from ufo2ft?
Are you building from a .glyphs source or from .designspace+UFOs?

ufo2ft featureWriters have two modes "skip" (default) and "append". In the skip mode, if the features.fea already contains the feature tag to be autogenerated, then nothing happens (which may be your case). In append mode, the newly autogenerated feautres are added at the end even if the same feature tag is already present in your feature file (which is what you'd like to do).

When building from .glyphs sources, fontmake (actually, glyphsLib) always sets "append" mode for the KernFeatureWriter so that it behaves accoriding to Glyphs.app's users expectations.

See https://github.com/googlefonts/glyphsLib/blob/0a6970a7b5b3c1e551dda5da08633afd872710fb/Lib/glyphsLib/builder/constants.py#L88-L102

If you are building from .designspace+UFO sources, one way to enable the "append" mode behavior for the KernFeatureWriter (which also builds the dist feature) is to add the following plist to the designspace "lib" element:

```xml
<designspace>

  <!-- ... -->

  <lib>
    <dict>
      <key>LWPluginVersion</key>
      <string>1429</string>
      <key>com.github.googlei18n.ufo2ft.featureWriters</key>
      <array>
        <dict>
          <key>class</key>
          <string>KernFeatureWriter</string>
          <key>options</key>
          <dict>
            <key>mode</key>
            <string>append</string>
          </dict>
        </dict>
        <dict>
          <key>class</key>
          <string>MarkFeatureWriter</string>
          <key>options</key>
          <dict>
            <key>mode</key>
            <string>skip</string>
          </dict>
        </dict>
      </array>
    </dict>
  </lib>
</designspace>
```

Which is what glyphsLib is doing here (using the fontTools.designspaceLib API):
https://github.com/googlefonts/glyphsLib/blob/0a6970a7b5b3c1e551dda5da08633afd872710fb/Lib/glyphsLib/builder/user_data.py#L53

Another way, if you are building using fontmake command line script, is passing the `--feature-writer` option. E.g.:

```
$ fontmake -m MyFont.designspace -o variable --feature-writer 'KernFeatureWriter(mode="append")' --feature-writer MarkFeatureWriter
```

Note that using `--feature-writer` option or set the `com.github.googlei18n.ufo2ft.featureWriters` lib key will override the default list of ufo2ft featureWriters, so if you only set the KernFeatureWriter (because you want to override the default mode), the mark/mkmk features will _not_ be generated unless you also include the MarkFeatureWriter in the list of writers to be enabled.

I hope that helps. Let me know if setting "append" mode does not solve your issue.

**Comment by rahulgajjar** [comment]
Hi @anthrotype 

I am using fontmake from command line with .designspace+UFO. I have little script to automate things. I tried with `KernFeatureWriter(mode="append")`
Now it is working as I intended.

Thank you so much.

Now lets hope other softwares implement this correctly. I am developing font for "Takri" script and it uses Universal Shaping Engine. I don't see double kerning issue with this font in Adobe Indesign even if both 'dist' and 'kern' blocks are present. (This did not work for 'Kannada' font) But now somehow Microsoft totally ignores 'dist' feature block
